Improve support for external request queues (#4783)
scrapy/scrapy: docs/topics/request-queues.rst

    affect request serialization speed (CPU usage) and size (disk usage). See
    “Comparison with ``marshal``” in the documentation of :mod:`pickle` for
    more information.

    When request serialization fails, Scrapy falls back to using a memory
    downstream queue.

    Scrapy provides the following disk downstream queue classes:

    .. autoclass:: MarshalFifoDiskQueue

    .. autoclass:: MarshalLifoDiskQueue

    .. autoclass:: PickleFifoDiskQueue

    .. autoclass:: PickleLifoDiskQueue

Writing your own disk downstream queue class
--------------------------------------------

If you want to define and use your own disk downstream queue implementation,
it has to conform to the following interface:

.. class:: MyExternalQueue

   .. classmethod:: from_crawler(cls, crawler: scrapy.crawler.Crawler, key: str, *args, **kwargs)

      Return an instance of this disk downstream queue class.

      *key* is the unique ID of the disk downstream queue instance. It may be
      used, for example, to create a unique file or folder to store the queue
      content.

      If the input data is invalid, raise an exception from this class method

>       of from your ``__init__`` method to halt the crawl.

Reviewer: We should describe the content and purpose of `*args` and `**kwargs` here.

Reviewer 2: it seems this comment is still valid

Author: Added a notion about queue internal state. Problem is storing/restoring is not supported in `ScrapyPriorityQueue`. Adding support isn't very hard, but there two reasons not implementing it right now: 
- it breaks on disk format
- implementing in backward-compatible way takes more time and postpone this PR even more
For future it is important to preserve it in protocol. In meantime a lot of good things are possible right now for queues not required storing/restoring.

What is your thoughts?

Author: I thought about it little more. I don't like current protocol for state passing. Let's me think more and came up with better solution.

Author: Hey @Gallaecio and @kmike I finished state passing protocol and implementation for `ScrapyPriorityQueue` with support for loading from obsolete format.

>    .. method:: push(self, request: scrapy.http.Request)

Reviewer 2: what's the return value? If it is None, it could make sense to annotate it as None

      Push a request into the queue.

      The helper function :func:`~scrapy.utils.reqser.request_to_dict` can be
      used to convert the request into a dict that can then be easily
      serialized with, for example, :func:`pickle.dumps`.

      Scrapy falls back to the memory downstream queue for *request* if one of
      the following exceptions is raised:

      -     :exc:`TransientError`: indicates a temporary failure.

            For example, if storing requests on a server, raise this exception
            if you temporarily loose access to the server.

        -   :exc:`SerializationError`: indicates that *request* could not be
            serialized

   .. method:: pop(self)

      Pop a request from the queue.

      In case of a temporary problem, ``None`` is returned. In all other cases,

>       an exception is raised, causing the crawling process to halt.

Reviewer 2: Would it make sense to annotate output type as `Optional[Request]`?

>       In case of a temporary problem, ``None`` is returned. In all other cases,
>       an exception is raised, causing the crawling process to halt.

Reviewer 2: ```suggestion
      In case of a temporary problem, ``None`` is returned.
      Exceptions raised in this method causes crawler to halt.
```
Though ^^ is still not super clear :) What does "crawling process to halt" means? Does it mean a spider gets shut down (but other spiders running in the same process can continue), that the Python process exits? If it is spider who shuts down, is it a hard shutdown or a soft one, i.e. are requests in Downloader still get processed?

>       in this method woudln't be handled and may lead to halt of the crawler.

Reviewer 3: ```suggestion
      in this method wouldn't be handled and may lead to halt of the crawler.
```

      The helper function :func:`~scrapy.utils.reqser.request_from_dict` can
      be used to convert a deserialized dict back into a

>       :class:`~scrapy.http.Request` object.

Reviewer 2: What should `pop` method return? Is it deserialized dict (and so the caller should call request_from_dict), or is it a Request object (and so this comment is a suggestion on how to implement `pop` method)?

Author: Typing is added and it's clear now

   .. method:: close(self)

      Release internal resources (e.g. close files or sockets).

   .. method:: __len__(self)

      Return the number of requests in the queue.

      If the number of requests cannot be determined (e.g. because of a
      connection problem), the method should neither return 0, which would
      cause the queue to be closed, nor raise an exception, which would halt

>       the crawl.

Reviewer 2: What should it do then? Should it return None, or 1, or a large number?

.. _topics-request-queues:

==============
Request queues
==============

.. _priority-queues:
.. setting:: SCHEDULER_PRIORITY_QUEUE

Priority queues
===============

.. currentmodule:: scrapy.pqueues

To decide the order in which requests are sent, Scrapy uses a queue called the
*priority queue*.

The :setting:`SCHEDULER_PRIORITY_QUEUE` setting determines the type of priority
queue that Scrapy uses. Its default value is :class:`ScrapyPriorityQueue`.

Scrapy provides the following priority queue implementations:

.. autoclass:: DownloaderAwarePriorityQueue

.. autoclass:: ScrapyPriorityQueue

.. setting:: SCHEDULER_DISK_QUEUE
.. setting:: SCHEDULER_MEMORY_QUEUE
.. _downstream-queues:

Downstream queues
=================

.. currentmodule:: scrapy.squeues

A :ref:`priority queue <priority-queues>` uses queues internally to store
pending requests. These internal queues are called *downstream queues*.

The organization method used by the downstream queues, usually either
FIFO or LIFO, affects the order in which requests are sent for requests
that the :ref:`priority queue <priority-queues>` considers to have the same
priority.

> The type of downstream queue used depends on the :setting:`JOBDIR` setting (see
> :ref:`topics-jobs`):

Reviewer: ```suggestion
The built-in :ref:`priority queues <priority-queues>` allow you to choose the 
type of downstream queue they use through a setting. Which setting, however, 
depends on the value of the :setting:`JOBDIR` setting (see :ref:`topics-jobs`):
```

    .. autoclass:: FifoMemoryQueue

    .. autoclass:: LifoMemoryQueue

-   If :setting:`JOBDIR` is set, the disk downstream queue from the
    :setting:`SCHEDULER_DISK_QUEUE` setting is used instead, which defaults to
    :class:`PickleLifoDiskQueue`.

    In addition to request order, which disk downstream queue you use can
    affect request serialization speed (CPU usage) and size (disk usage). See
    “Comparison with ``marshal``” in the documentation of :mod:`pickle` for
    more information.

    When request serialization fails, Scrapy falls back to using a memory
    downstream queue.

    Scrapy provides the following disk downstream queue classes:

    .. autoclass:: MarshalFifoDiskQueue

    .. autoclass:: MarshalLifoDiskQueue

    .. autoclass:: PickleFifoDiskQueue

    .. autoclass:: PickleLifoDiskQueue

Writing your own disk downstream queue class
--------------------------------------------

If you want to define and use your own disk downstream queue implementation,
it has to conform to the following interface:

.. class:: MyExternalQueue

   .. classmethod:: __init__(cls, crawler: scrapy.crawler.Crawler, key: str, state)

>       Return an instance of this disk downstream queue class.

Reviewer: `__init__` is not a `@classmethod`, and it does not “return” anything.

Also, these parameters seems to be based on the custom queue class having defined a `from_crawler` method, which it does not need to define. We should cover `from_crawler` in this class documentation as an optional classmethod, similar to how it’s covered for other Scrapy components (e.g. item pipelines). But the documentation of this `__init__` method cannot assume a `from_crawler` is defined, and so it cannot assume that the `crawler` parameter is received.

Author: > Also, these parameters seems to be based on the custom queue class having defined a from_crawler method

`crawler` argument is used by downstream queues to get settings etc.  

> and so it cannot assume that the crawler parameter is received.

Here I didn't understand why we cannot assume argument of function isn't passed?

Author: > __init__ is not a @classmethod

Which mark should we use here for proper formatting?

Reviewer: > `crawler` argument is used by downstream queues to get settings etc. […] Here I didn't understand why we cannot assume argument of function isn't passed?

Oh, I see now.

Usually Scrapy components do not get a crawler directly, but they are created with `create_instance` which calls the class method `from_crawler` if defined, passing the `crawler` object and all other parameters to that class method, which then calls `__init__`. I wrongly assumed that was the case here as well.

In this case, it looks like (for historic reasons?) the `crawler` object is being passed directly on the call to `create_instance`, so there is no need to define `from_crawler` to get the `crawler` object:

https://github.com/scrapy/scrapy/blob/68379197986ae3deb81a545b5fd6920ea3347094/scrapy/pqueues.py#L75-L81

So, please forget the second part of my comment completely.

> Which mark should we use here for proper formatting?

I believe something like this (classmethod → method, cls → self):

```rst
.. method:: __init__(self, crawler: scrapy.crawler.Crawler, key: str, state)
```

Author: Sorry, this is very old PR and I relied on a memory. Actually you are right. We need to pass a crawler we need `from_crawler` method here, as https://github.com/scrapy/scrapy/blob/0915be7300b19e9f2df2957734e2710a709bb086/scrapy/utils/misc.py#L171-L175

Author: `__init__` is replaced by `from_crawler` in the documentation

    :class:`PickleLifoDiskQueue`.

    In addition to request order, which disk downstream queue you use can
    affect request serialization speed (CPU usage) and size (disk usage). See
    “Comparison with ``marshal``” in the documentation of :mod:`pickle` for
    more information.

    When request serialization fails, Scrapy falls back to using a memory
    downstream queue.

    Scrapy provides the following disk downstream queue classes:

    .. autoclass:: MarshalFifoDiskQueue

    .. autoclass:: MarshalLifoDiskQueue

    .. autoclass:: PickleFifoDiskQueue

    .. autoclass:: PickleLifoDiskQueue

Writing your own disk downstream queue class
--------------------------------------------

If you want to define and use your own disk downstream queue implementation,
it has to conform to the following interface:

.. class:: MyExternalQueue

   .. classmethod:: __init__(cls, crawler: scrapy.crawler.Crawler, key: str, state)

      Return an instance of this disk downstream queue class.

      *key* is the unique ID of the disk downstream queue instance. It may be
      used, for example, to create a unique file or folder to store the queue
      content.

>       *arg* and *kwargs* are used to restore internal state.

Reviewer 4: They are no longer included in the signature, should this be removed?

Author: fixed

>       *state* is a variable containing state of queue. It should be serializeable
>       and deserializeble by json module. You need to come up with proper type
>       and structure to completely restore an internal state of your queue.

Reviewer: Thanks, I think I now understand how this works from just the documentation.

What would you think about a bit of rephrasing?

```suggestion
      *state* is the data returned by the :method:`close` method of this queue 
      during a prior execution, and should be used to restore the queue to that 
      prior state. If there is no prior execution, its value is ``None``.
```

With a reference to `close` here now, I think it makes sense to keep the rest 
of the information in the documentation of the `close` method that is 
responsible for creating the state data.

   .. method:: push(self, request: scrapy.http.Request) -> None

      Push a request into the queue.

      The helper function :func:`~scrapy.utils.reqser.request_to_dict` can be
      used to convert the request into a dict that can then be easily
      serialized with, for example, :func:`pickle.dumps`.

      Scrapy falls back to the memory downstream queue for *request* if one of
      the following exceptions is raised:

      -     :exc:`TransientError`: indicates a temporary failure.

            For example, if storing requests on a server, raise this exception

>             if you temporarily loose access to the server.

Reviewer: ```suggestion
            if you temporarily lose access to the server.
```

        -   :exc:`SerializationError`: indicates that *request* could not be

>             serialized

Reviewer: ```suggestion
            serialized.
```

   .. method:: pop(self) -> Optional[Request]

      Pop a request from the queue.

      In case of a temporary problem or empty queue, ``None`` is returned.
      Exceptions raised in this method woudln't be handled and may lead to

>       halt of the crawler.

Reviewer: ```suggestion
      a halt of the crawler.
```

      The helper function :func:`~scrapy.utils.reqser.request_from_dict` can
      be used to convert a deserialized dict back into a
      :class:`~scrapy.http.Request` object.

   .. method:: peek(self) -> Optional[Request]

>       Returns the next object to be returned by :meth:`pop`,
>       but without removing it from the queue.
> 
>       Raises :exc:`NotImplementedError` if the underlying queue class does
>       not implement a ``peek`` method, which is optional for queues.
> 
>       In case of a temporary problem or empty queue, ``None`` is returned.

Reviewer: I’ve just realized that we are lying in the documentation of `ScrapyPriorityQueue.peek` currently in `master`, where we say the same thing about `NotImplementedError`. Our built-in queues just happen to raise `NotImplementedError`, but if a user does not define this method `AttributeError` will be raised instead.

I’m not sure how to best approach this, though. My current thinking is:
- Allow users not to define this method at all, and update that second paragraph above accordingly.
- Have our built-in queues without peek support either not define the method at all or raise `AttributeError` instead of `NotImplementedError`.
- Update the documentation of our public API affected by this change, such as `ScrapyPriorityQueue.peek`.
- This is technically a backward-incompatible change, but I believe the affected classes (directly and indirectly) did not have public documentation before, so maybe it could be argued that raising `NotImplementedError` was an implementation detail?

On a different note, we should use the imperative form here:

```suggestion
      Return the next object to be returned by :meth:`pop`,
      but without removing it from the queue.

      Raise :exc:`NotImplementedError` if the underlying queue class does
      not implement a ``peek`` method, which is optional for queues.

      In case of a temporary problem or empty queue, return ``None``.
```

   .. method:: close(self) -> Optional[Any]

>       Release internal resources (e.g. close files or sockets) and return
>       current internal state if any. This state will be seriliazed and passed
>       to __init__ method.

Reviewer: Here we can include the detail about the data needing to be JSON-serializable, as well as a reference to the topic about pausing and resuming crawls, for better context.

```suggestion
      Release internal resources (e.g. close files or sockets) and return an 
      object serializable by :function:`json.dump` (e.g. a :class:`dict`) that 
      represents the current internal state of this queue. When :ref:`resuming 
      a crawl <topics-jobs>`, this object is passed to the :method:`__init__` 
      method of this queue class to restore the queue state.
```

    defaults to :class:`LifoMemoryQueue`.

    Scrapy provides the following memory downstream queue classes:

    .. autoclass:: FifoMemoryQueue

    .. autoclass:: LifoMemoryQueue

-   If :setting:`JOBDIR` is set, the disk downstream queue from the
    :setting:`SCHEDULER_DISK_QUEUE` setting is used instead, which defaults to
    :class:`PickleLifoDiskQueue`.

    In addition to request order, which disk downstream queue you use can
    affect request serialization speed (CPU usage) and size (disk usage). See
    “Comparison with ``marshal``” in the documentation of :mod:`pickle` for
    more information.

    When request serialization fails, Scrapy falls back to using a memory
    downstream queue.

    Scrapy provides the following disk downstream queue classes:

    .. autoclass:: MarshalFifoDiskQueue

    .. autoclass:: MarshalLifoDiskQueue

    .. autoclass:: PickleFifoDiskQueue

    .. autoclass:: PickleLifoDiskQueue

Writing your own disk downstream queue class
--------------------------------------------

If you want to define and use your own downstream queue implementation,
it has to conform to the following interface:

.. class:: MyExternalQueue

>    .. method:: __init__(self, crawler: scrapy.crawler.Crawler, downstream_queue_cls: Optional[T], key: str, state)

Reviewer: `MyExternalQueue` _is_ `downstream_queue_cls`, right?. I think `downstream_queue_cls` is only passed to priority queues, the parameter should not exist here.

>    .. classmethod:: from_crawler(cls, downstream_queue_cls: Optional[T], crawler: scrapy.crawler.Crawler, key: str, state)

Author: I removed `from_crawler` usage for queues and transition to contructor usage everywhere

>    .. classmethod:: from_crawler(cls, downstream_queue_cls: Optional[T], crawler: scrapy.crawler.Crawler, key: str, state)

Reviewer: Why? The general tendency is to try and offer `from_crawler` support in all Scrapy components. And if you are going to pass the crawler anyway, you may as well use `from_crawler` to do it. If nothing else, supporting `from_crawler` simplifies the signature for custom queues that do not need `crawler` access.

      Initializes an instance of this disk downstream queue class.

      *downstream_queue_cls* class of queues lying down on queue hierarchy, e.g.
      priority queue manages disk queues. ``None`` in case there is no downstream
      queue

      *key* is the unique ID of the disk downstream queue instance. It may be
      used, for example, to create a unique file or folder to store the queue
      content.

>       *state* is the data returned by the :method:`close` method of this queue

Reviewer: My bad, it’s:

```suggestion
      *state* is the data returned by the :meth:`close` method of this queue
```

https://www.sphinx-doc.org/en/master/usage/restructuredtext/domains.html#role-py-meth

      during a prior execution, and should be used to restore the queue to that
      prior state. If there is no prior execution, its value is ``None``.

   .. method:: push(self, request: scrapy.http.Request) -> None

      Push a request into the queue.

      The helper function :func:`~scrapy.utils.reqser.request_to_dict` can be
      used to convert the request into a dict that can then be easily
      serialized with, for example, :func:`pickle.dumps`.

      Scrapy falls back to the memory downstream queue for *request* if one of
      the following exceptions is raised:

      -     :exc:`TransientError`: indicates a temporary failure.

            For example, if storing requests on a server, raise this exception
            if you temporarily lose access to the server.

        -   :exc:`SerializationError`: indicates that *request* could not be
            serialized.

   .. method:: pop(self) -> Optional[Request]

      Pop a request from the queue.

>       In case of a temporary problem or empty queue, ``None`` is returned.

Reviewer: :lipstick: 

```suggestion
      In case of a temporary problem or empty queue, return ``None``.
```

      Exceptions raised in this method woudln't be handled and may lead to
      a halt of the crawler.

      The helper function :func:`~scrapy.utils.reqser.request_from_dict` can
      be used to convert a deserialized dict back into a
      :class:`~scrapy.http.Request` object.

   .. method:: peek(self) -> Optional[Request]

      Return the next object to be returned by :meth:`pop`,
      but without removing it from the queue.

>       Raise :exc:`AttributeError` if the underlying queue class does
>       not implement a ``peek`` method, which is optional for queues.

Reviewer: ```suggestion
      Implementing this method is optional. If implemented, but peek support
      is not possible, raise :exc:`AttributeError`.
```
